Clarify SourceLookup sharing across fetch subphases.

jtibshirani opened the pull request:
The `SourceLookup` class provides access to the _source for a particular
document, specified through `SourceLookup#setSegmentAndDocument`. Previously
the search context contained a single `SourceLookup` that was shared between
different fetch subphases. It was hard to reason about its state: is
`SourceLookup` set to the expected document? Is the _source already loaded and
available?

Instead of using a global source lookup, the fetch hit context now provides
access to a lookup that is set to load from the hit document.

This refactor closes #31000, since the same `SourceLookup` is no longer shared
between the 'fetch _source phase' and script execution.

jtibshirani committed:
Create new SourceLookup for SignificantTextAggregator.

jtibshirani committed:
Store SourceLookup on HitContext and set the source when creating hits.

jtibshirani committed:
In highlighting, access source through HitContext.

jtibshirani committed:
Stop saving parent _source before loading inner hits.

This saved _source actually never appears to be used.

jtibshirani committed:
Remove SearchContext#lookup.

jtibshirani committed:
Fix null pointer in FetchSourcePhaseTests.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsPhase.java:
```diff
@@ -65,10 +65,6 @@ public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOExcept
                 }
                 innerHits.docIdsToLoad(docIdsToLoad, 0, docIdsToLoad.length);
                 innerHits.setId(hit.getId());
-                innerHits.lookup().source().setSource(context.lookup().source().internalSourceRef());
```
We set the root document's _source here, but when processing inner hits we didn't use it and instead reloaded the _source. So this is safe to remove for now.

In a follow-up PR, I will fix this to make sure the _source is loaded only once for the root doc + inner hits.

jimczi replied:
+1, that's something that we spotted sometime ago: https://github.com/elastic/elasticsearch/issues/32818#issuecomment-412878390 and maybe https://github.com/elastic/elasticsearch/issues/56210 but never implemented.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/fetch/FetchSubPhase.java:
```diff
@@ -39,13 +40,15 @@
         private IndexSearcher searcher;
         private LeafReaderContext readerContext;
         private int docId;
+        private final SourceLookup sourceLookup = new SourceLookup();
```
Do we always need to create `SourceLookup` object here?  What about cases where we don't need _source?

jtibshirani replied:
To me the optimization might not be worth it, it's not an expensive object to construct and we only create one `SourceLookup` per hit?

mayya-sharipova commented:
@jtibshirani Thanks for the PR, I like the idea. I had an initial look, and will look more thoroughly tomorrow.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java:
```diff
@@ -154,13 +153,12 @@ protected Aggregator createInternal(SearchContext searchContext, Aggregator pare
             boolean filterDuplicateText
         ) {
-            this.sourceLookup = sourceLookup;
+            this.sourceLookup = new SourceLookup();
```
Since this one happens at query time maybe we can keep using it? I think what you have here would cause two significant text aggs to load source twice.

jtibshirani replied:
Oops, this is a great point. I think I took it the refactor too far. I plan to restore the shared `SourceLookup` on `SearchLookup` so it can be used to share cached _source in the query phase.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java:
```diff
@@ -27,31 +27,22 @@
-                sourceLookup,
-                fieldsLookup.getLeafFieldsLookup(context));
+                docLookup.getLeafDocLookup(context),
+                new SourceLookup(),
```
Similar to @nik9000's comment, it looks like source will be loaded twice here for cases when a script needs to access ` _source` and when `_source` is requested for a document, for example for queries like this:

```js
{
  "_source": true,
  "query": {
    "match_all": {}
  },
  "script_fields": {
    "my_script": {
      "script": {
        "source": "params['_source']['my_field']"
      }
    }
  }
}
```

jtibshirani replied:
I'm not sure this is the case -- looking at `ScriptFieldsPhase`, it overwrites `hitsExecute` and so processes all hits at once. That means it wouldn't be able to re-use the loaded _source (even before the refactor) ?

jtibshirani replied:
Another note: I'll add some Javadoc to `HitContext` explaining the _source sharing strategy, since it is quite tricky to understand.

nik9000 replied:
I do think this'll break the shared loading of `_source` in scripts at query time though. Each script calls this method to get its own fresh LeafSearchLookup and we sort of rely on them being able to share the loaded `_source`.

jtibshirani replied:
> I do think this'll break the shared loading of _source in scripts at query time though.

👍  I agree, this will be fixed when I restore the shared `SourceLookup` on `SearchLookup`.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/FetchSubPhase.java:
```diff
@@ -39,13 +40,15 @@
         private IndexSearcher searcher;
         private LeafReaderContext readerContext;
         private int docId;
+        private final SourceLookup sourceLookup = new SourceLookup();
```
I think we should add a comment here about why we don't want to share the lookup in QueryShardContext.

jtibshirani replied:
👍

jtibshirani replied:
I added Javadoc on `HitContext#sourceLookup` and `FetchPhase#prepareHitContext` to explain the logic. Let me know if it could use more clarification.

nik9000 commented on server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java:
```diff
@@ -27,31 +27,22 @@
-
-    final DocLookup docMap;
-
-    final SourceLookup sourceLookup;
```
I think we still need `SourceLookup` here both for significant text, and for the thing we talked about where I wanted `SearchLookup` on `FieldMapper#lookup`.

jtibshirani replied:
👍  will restore it.

jtibshirani committed:
Restore shared SourceLookup to SearchLookup.

jtibshirani commented:
I caught up with @nik9000 offline about the refactor and wanted to add more context as to its benefits:
* The _source loading semantics are clearer. We attempt to load the _source once at the start of the fetch phase, and explicitly add it to the `HitContext` to make it available for the subphases. Before we manipulated global state (setting the _source on `SearchContext`), and it was not obvious what the intention was.
* Any subphase that overrides `hitExecute` (singular) will have access to the hit context, which contains a reference to the _source for the current document. The subphases that override `hitsExecute` do not have access to a hit context and need to reload _source. Previously this difference in behavior was unclear.

However I now see a downside to this refactor. There used to be only one way to share a `SourceLookup` instance (as part of `SearchLookup` on `QueryShardContext`), but now we've added another specific to the fetch phase. This means that a subphase that overwrites `hitExecute` and also uses a script will not share the loaded _source.

I'm curious as to what others think. I still like the refactor and feel that it moves us in a better direction.

jimczi commented:
> However I now see a downside to this refactor. There used to be only one way to share a SourceLookup instance (as part of SearchLookup on QueryShardContext), but now we've added another specific to the fetch phase. This means that a subphase that overwrites hitExecute and also uses a script will not share the loaded _source.

We don't have such example at the moment, right ? That's also something we can tackle in a follow up, currently we use the lookup of the query shard context to compile the scripted field but we could refactor this part to move to the new hit context ? Hits are now [sorted by doc ids](https://github.com/elastic/elasticsearch/pull/57273) in fetch sub phases so it's also a good chance to get ride of the `hitsExecute` completely. That would allow to load the `_source` only once since today sub phases that use `hitsExecute` have no way to reuse `_source` ?

jtibshirani committed:
Add comments explaining source loading + sharing.

jtibshirani commented:
> We don't have such example at the moment, right ?

I also can't think of an example right now.

> ... so it's also a good chance to get ride of the hitsExecute completely.

This is a good goal, and like you said I think the refactor leaves open these future simplifications.

jtibshirani commented:
> We don't have such example at the moment, right ?
> I also can't think of an example right now.

A correction here: I hadn't realized that `_source` was available within score scripts. So it is possible to have an 'explain' phase for a query that uses a score script (that then accesses `_source`). I'm not really worried about optimizing performance in this case though, since 'explain' is intended as a debugging tool.

jimczi commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -222,42 +221,60 @@ private int findRootDocumentIfNested(SearchContext context, LeafReaderContext su
-        SourceLookup sourceLookup = context.lookup().source();
-        sourceLookup.setSegmentAndDocument(subReaderContext, subDocId);
-        if (fieldsVisitor.source() != null) {
-            sourceLookup.setSource(fieldsVisitor.source());
```
I wonder if we can still assign the `context.lookup().source()` with the current _source so that `_script` can share the source with other fetch sub-phase. That would only work for sub-phases that use a script and override `hitExecute` so the new `FetchFieldsPhase` would benefit from this hack ? We can then think how we could clenup the context for script in a follow up ?

jtibshirani replied:
I don't think `FetchFieldsPhase` would benefit currently, because it doesn't support scripts? I like the plan of considering it in a follow-up, for example as part of converting `ScriptFieldsPhase` to use `hitExecute`.

jimczi replied:
Right, I was thinking preemptively of runtime scripted field but +1 to consider this in a follow up.

jimczi commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/SourceScoreOrderFragmentsBuilder.java:
```diff
@@ -36,11 +35,10 @@
 
     public SourceScoreOrderFragmentsBuilder(MappedFieldType fieldType,
-                                            QueryShardContext context,
+                                            FetchSubPhase.HitContext context,
```
Maybe just pass the `SourceLookup` directly ?

jtibshirani replied:
👍

jimczi commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/SourceSimpleFragmentsBuilder.java:
```diff
@@ -21,21 +21,20 @@
 
     public SourceSimpleFragmentsBuilder(MappedFieldType fieldType,
-                                        QueryShardContext context,
+                                        FetchSubPhase.HitContext context,
```
Same here ?

jtibshirani replied:
👍

jimczi reviewed (approved):
I like this change. I left one comment regarding a possible hack to ensure that scripts and fetch sub phase can still share the _source but I don't feel strongly about it. We can solve the double load more nicely in a follow up if needed.

jtibshirani committed:
Undo all changes to SearchLookup class.

They are not closely related to this PR and it's clearer to keep them separate.

jtibshirani committed:
Avoid passing whole HitContext to the highlight fragments builders.

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into source-lookup

jtibshirani commented:
In addition to restoring the shared `SourceLookup` on `SearchLookup`, I ended up removing all changes to the `SearchLookup` class, as they weren't core to this PR. So now the PR is scoped just to 'fetch' and avoids @nik9000 and @mayya-sharipova's concern around scripts sharing _source in the query phase.

Here is the final summary of changes:
* Store `SourceLookup` on `HitContext` and set the source when creating hits.
* Fetch subphases access source through `HitContext` when possible.
* Stop saving parent _source before loading inner hits (it was unused).
* Remove `SearchContext#lookup` method.

jtibshirani committed:
Remove unnecessary method mock in AggregatorTestCase.

jtibshirani committed:
Fix checkstyle violation.

jtibshirani committed:
Undo the change to SourceLookup sharing in SIgnificantTextAggregator.

jtibshirani commented:
@elasticmachine run elasticsearch-ci/bwc

mayya-sharipova reviewed (approved):
@jtibshirani Thanks Julie, looks a very good change to me!

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into source-lookup

nik9000 reviewed (approved):
LGTM2

jtibshirani commented:
Thanks everyone, the review feedback was really helpful.
